### PR TITLE
feat(metrics): Propagate flow_* params from content to payments server

### PR DESCRIPTION
**Pull request by shane-tomlinson**
This fetches the flow_begin_time and flow_id out of a view's metrics and propagates them whenever redirecting to the payments server as the flow_begin_time and flow_id query params.

There were some indentation fixes, so this might be easier to review with ?w=1

issue #2425
issue #2801

@6a68, @chenba - r?

### CODE REVIEW

File: packages/fxa-content-server/app/scripts/lib/payment-server.js [hunk 1]
```diff
@@ -4,38 +4,54 @@
...
-    queryParams
+    queryParams = {}
   ) {
     const {
       managementClientId,
       managementScopes,
       managementTokenTTL,
       managementUrl,
     } = subscriptionsConfig;
-    const queryString =
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/app/scripts/lib/payment-server.js]: The Url module does this, so I removed the duplicate logic here.

File: packages/fxa-content-server/app/scripts/lib/url.js [hunk 2]
```diff
@@ -114,7 +114,11 @@ export default {
     const params = [];
     for (const paramName in obj) {
       const paramValue = obj[paramName];
-      if (typeof paramValue !== 'undefined' && paramValue !== null) {
+      if (
+        typeof paramValue !== 'undefined' &&
+        paramValue !== null &&
+        paramValue !== ''
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/app/scripts/lib/url.js]: I did have to update this to filter out empty param values though.

File: packages/fxa-content-server/app/tests/spec/lib/payment-server.js [hunk 3]
```diff
@@ -32,12 +32,18 @@ describe('lib/payment-server-redirect', () => {
...
+          flowId: 'foo',
+        })),
+      },
     };
   });
 
   it('redirects as expected', () => {
     const REDIRECT_PATH = 'example/path';
-    PaymentServer.navigateToPaymentServer(
+    return PaymentServer.navigateToPaymentServer(
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/app/tests/spec/lib/payment-server.js]: These tests weren't returning the promises and were actually failing, so had to do a bit of tangential repair.

File: packages/fxa-content-server/app/tests/spec/views/subscriptions_management_redirect.js [hunk 4]
```diff
@@ -2,78 +2,75 @@
...
-
-     it('calls PaymentServer.navigateToPaymentServer as expected', () => {
-       assert.deepEqual(PaymentServer.navigateToPaymentServer.args, [
-         [view, config.subscriptions, `subscriptions`],
-       ]);
-     });
-   });
- });
+  describe('render', () => {
+    it('renders correctly, initializes flow evnets, navigates to payments server', () => {
```
Comment (6a68) [inline packages/fxa-content-server/app/tests/spec/views/subscriptions_management_redirect.js]: "evnets" -> "events"

File: packages/fxa-content-server/app/scripts/views/mixins/flow-events-mixin.js [hunk 5]
```diff
@@ -22,6 +22,13 @@ export default {
     submit: '_submitFlowEventsForm',
   },
 
+  /**
+   * Initialize the flow events
```
Comment (6a68) [inline packages/fxa-content-server/app/scripts/views/mixins/flow-events-mixin.js]: Not sure this is a useful comment, I'd probably omit it

File: packages/fxa-content-server/app/tests/spec/lib/payment-server.js (hunk 3 above)
Comment (lmorchard) [reply packages/fxa-content-server/app/tests/spec/lib/payment-server.js]: (whoops!)

File: packages/fxa-content-server/app/scripts/lib/payment-server.js [hunk 6]
```diff
@@ -4,38 +4,54 @@
...
-      Object.keys(queryParams)
-        .filter(k => queryParams[k] !== null && queryParams[k] !== '')
-        .map(
-          k => `${encodeURIComponent(k)}=${encodeURIComponent(queryParams[k])}`
-        )
-        .join('&');
-    const qs = queryString.length ? `?${queryString}` : '';
+
+    const { flowBeginTime, flowId } = view.metrics.getFlowEventMetadata();
+
```
Comment (chenba) [inline packages/fxa-content-server/app/scripts/lib/payment-server.js]: FWIW I see the in teh logs that `device_id` is used for an amplitude event, but not the flowBeginTime.

Comment (shane-tomlinson) [reply packages/fxa-content-server/app/scripts/lib/payment-server.js]: I'll add that.

File: packages/fxa-content-server/app/scripts/lib/payment-server.js [hunk 7]
```diff
@@ -4,38 +4,59 @@
...
-          accessToken.get('token')
-        )}`;
+        const queryString = Url.objToSearchString({
+          // device_id, flow_begin_time, and flow_id need to be propagated to
+          // the payments server so that the user funnel can be traced from the RP,
+          // through the content server, and to the payments server, appearing as
+          // the same user throughout.
+          device_id: deviceId,
+          flow_begin_time: flowBeginTime,
+          flow_id: flowId,
```
Comment (chenba) [inline packages/fxa-content-server/app/scripts/lib/payment-server.js]: Can these be in camelCase?  I don't know the reason for the snake_case here.

Comment (chenba) [reply packages/fxa-content-server/app/scripts/lib/payment-server.js]: Oh I just noticed that it's been merged.  No prob. :D

**Pull request merged**